src/StorageRent.sol

```diff
@@ -0,0 +1,626 @@
+        if (priceAnsweredInRound < priceRoundId) revert StaleAnswer();
+
+        /* Set the last update timestamp and block. */
+        lastPriceFeedUpdateTime = block.timestamp;
+        lastPriceFeedUpdateBlock = block.number;
+
+        if (prevEthUsdPrice == 0 && ethUsdPrice == 0) {
+            /* If this is the very first price update, set previous equal to latest. */
+            prevEthUsdPrice = ethUsdPrice = uint256(answer);
+        } else {
+            prevEthUsdPrice = ethUsdPrice;
+            ethUsdPrice = uint256(answer);
+        }
+    }
+
+    function _price(uint256 units) internal returns (uint256) {
+        return _price(units, usdUnitPrice, _ethUsdPrice());
+    }
+
+    /**
+     * @param units      Number of storage units. Integer, no decimals.
+     * @param usdPerUnit Unit price in USD. Fixed point with 8 decimals.
+     * @param ethPerUsd  ETH/USD price. Fixed point with 8 decimals.
+     *
+     * @return uint256 price in wei, i.e. 18 decimals.
+     */
+    function _price(uint256 units, uint256 usdPerUnit, uint256 ethPerUsd) internal pure returns (uint256) {
+        return (units * usdPerUnit).divWadUp(ethPerUsd);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                         PERMISSIONED ACTIONS
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Credit a single fid with free storage units. Only callable by operator.
+     *
+     * @param fid   The fid that will receive the credit.
+     * @param units Number of storage units to credit.
+     */
+    function credit(uint256 fid, uint256 units) external onlyOperator whenNotDeprecated {
+        if (units == 0) revert InvalidAmount();
+        if (rentedUnits + units > maxUnits) revert ExceedsCapacity();
+
+        rentedUnits += units;
+        emit Rent(msg.sender, fid, units);
+    }
+
+    /**
+     * @notice Credit multiple fids with free storage units. Only callable by operator.
+     *
+     * @param fids  An array of fids.
+     * @param units Number of storage units per fid.
+     */
+    function batchCredit(uint256[] calldata fids, uint256 units) external onlyOperator whenNotDeprecated {
+        if (units == 0) revert InvalidAmount();
+        uint256 totalUnits = fids.length * units;
+        if (rentedUnits + totalUnits > maxUnits) revert ExceedsCapacity();
+        rentedUnits += totalUnits;
+        for (uint256 i; i < fids.length; ++i) {
+            emit Rent(msg.sender, fids[i], units);
+        }
+    }
+
+    /**
+     * @notice Credit a continuous sequence of fids with free storage units. Only callable by operator.
+     *
+     * @param start Lowest fid in sequence (inclusive).
+     * @param end   Highest fid in sequence (inclusive).
+     * @param units Number of storage units per fid.
+     */
+    function continuousCredit(uint256 start, uint256 end, uint256 units) external onlyOperator whenNotDeprecated {
```

Comment:
if we won't ever realistically use this after the initial seed, should we just always have start == 1? that simplifies this logic even further